### PR TITLE
ZK-3871: shadow components inside a Listbox doesn't work in nested way

### PR DESCRIPTION
**Pull request by klyvechen**

### CODE REVIEW

File: zktest/src/archive/test2/config.properties [hunk 1]
```diff
@@ -2414,6 +2414,7 @@ B85-ZK-3810.zul=A,E,Locale,WebSocket,i18n
 B85-ZK-3743.zul=A,E,Servlet3,AsyncContext,complete,CometServerPush
 B85-ZK-3804.zul=A,E,MeshWidget,Column,Width,Visible
 ##zats##B85-ZK-3866.zul=A,E,radio,doClick,evt,stop,popup
+##zats##B85-ZK-3871.zul=A,E,shadowElement,listbox,grid,nested
```
Comment (scribetw) [inline zktest/src/archive/test2/config.properties]: There is no `B85-ZK-3871.zul`.

Add 3 files:

- B85-ZK-3871-if.zul
- B85-ZK-3871-chooseif.zul
- B85-ZK-3871-choose.zul (Please include this file)

File: zktest/src/org/zkoss/zktest/test2/B85_ZK_shadowTestVM.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+        model.add(new InnerModel(true, "item3"));
+        model.add(new InnerModel(false, "item4"));
+        model.add(new InnerModel(true, "item5"));
+        //model1.add("item1*");
+        
+    }
+    
+    public ListModelList<InnerModel> model = new ListModelList();
+    
+    public ListModelList<InnerModel> model1 = new ListModelList();
```
Comment (scribetw) [inline zktest/src/org/zkoss/zktest/test2/B85_ZK_shadowTestVM.java]: No used. Just delete it

File: zktest/src/org/zkoss/zktest/test2/B85_ZK_shadowTestVM.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+ 
+ Copyright (C) 2018 Potix Corporation. All Rights Reserved.
+ */
+package org.zkoss.zktest.test2;
+
+import org.zkoss.bind.annotation.Command;
+import org.zkoss.bind.annotation.NotifyChange;
+import org.zkoss.zul.ListModelList;
+
+public class B85_ZK_shadowTestVM {
```
Comment (scribetw) [inline zktest/src/org/zkoss/zktest/test2/B85_ZK_shadowTestVM.java]: B85_ZK_3871VM

File: zk/src/org/zkoss/zk/ui/sys/ShadowElementsCtrl.java [hunk 4]
```diff
@@ -128,4 +120,20 @@ public static final Object getDistributedIndexInfo() {
 		}
 		return shadows;
 	}
+
+	private static void doBindChildrenInFilter(Component[] shadows) {
+		for (Component shadow : shadows) {
+			ShadowElement se = (ShadowElement) shadow;
+			if (!(shadow instanceof ShadowElement) || !se.getDistributedChildren().isEmpty() ||
+					!((ShadowElementCtrl) se).isDynamicValue())
+				return;
```
Comment (wenninghsu) [inline zk/src/org/zkoss/zk/ui/sys/ShadowElementsCtrl.java]: should be continue

**Pull request merged**